server/src/main/java/org/opensearch/index/query/BaseQueryRewriteContext.java

```diff
@@ -0,0 +1,140 @@
+    private final NamedXContentRegistry xContentRegistry;
+    private final NamedWriteableRegistry writeableRegistry;
+    protected final Client client;
+    protected final LongSupplier nowInMillis;
+    private final List<BiConsumer<Client, ActionListener<?>>> asyncActions = new ArrayList<>();
+    private final boolean validate;
+
+    public BaseQueryRewriteContext(
+        NamedXContentRegistry xContentRegistry,
+        NamedWriteableRegistry writeableRegistry,
+        Client client,
+        LongSupplier nowInMillis
+    ) {
+        this(xContentRegistry, writeableRegistry, client, nowInMillis, false);
+    }
+
+    public BaseQueryRewriteContext(
+        NamedXContentRegistry xContentRegistry,
+        NamedWriteableRegistry writeableRegistry,
+        Client client,
+        LongSupplier nowInMillis,
+        boolean validate
+    ) {
+
+        this.xContentRegistry = xContentRegistry;
+        this.writeableRegistry = writeableRegistry;
+        this.client = client;
+        this.nowInMillis = nowInMillis;
+        this.validate = validate;
+    }
+
+    /**
+     * The registry used to build new {@link XContentParser}s. Contains registered named parsers needed to parse the query.
+     */
+    public NamedXContentRegistry getXContentRegistry() {
+        return xContentRegistry;
+    }
+
+    /**
+     * Returns the time in milliseconds that is shared across all resources involved. Even across shards and nodes.
+     */
+    public long nowInMillis() {
+        return nowInMillis.getAsLong();
+    }
+
+    public NamedWriteableRegistry getWriteableRegistry() {
+        return writeableRegistry;
+    }
+
+    /**
+     * Returns an instance of {@link QueryShardContext} if available of null otherwise
+     */
+    public QueryShardContext convertToShardContext() {
+        return null;
+    }
+
+    /**
+     * Registers an async action that must be executed before the next rewrite round in order to make progress.
+     * This should be used if a rewriteabel needs to fetch some external resources in order to be executed ie. a document
+     * from an index.
+     */
+    public void registerAsyncAction(BiConsumer<Client, ActionListener<?>> asyncAction) {
+        asyncActions.add(asyncAction);
+    }
+
+    /**
+     * Returns <code>true</code> if there are any registered async actions.
+     */
+    public boolean hasAsyncActions() {
+        return asyncActions.isEmpty() == false;
+    }
+
+    /**
+     * Executes all registered async actions and notifies the listener once it's done. The value that is passed to the listener is always
+     * <code>null</code>. The list of registered actions is cleared once this method returns.
+     */
+    public void executeAsyncActions(ActionListener listener) {
+        if (asyncActions.isEmpty()) {
+            listener.onResponse(null);
+            return;
+        }
+
+        CountDown countDown = new CountDown(asyncActions.size());
```

Comment:
This countdown also has a race condition in that you set it prior to copying the list so could miss new additions.  This should probably be moved after the suggested synchronized block below, and count on the "new" list you have control over.

Reply:
this is what QueryRewriteContext has initially in [here](https://github.com/opensearch-project/OpenSearch/blob/abb81120ab4cc9b61e91f5b57a797409a951e3d7/server/src/main/java/org/opensearch/index/query/QueryRewriteContext.java#L134) , I just tried to create a BaseQueryRewriteContext to replace the original QueryRewriteContext. If we want to refactor the original QueryRewriteContext, we might want to do it in a separate PR.

Reply:
I write buggy code all the time, and I'm happy when someone comes behind me and fixes it.

Either this method is already protected from multithreading (in which case the copy to avoid concurrent modification seems silly) or there's a legitimate multithreading concern here that we should fix as long as we're touching the code.  

I agree with the earlier conversation that a `CopyOnWriteArray` is overkill, but copying in this case is pretty much useless if you don't synchronize the creation of the copy and clearing of the old list.

Reply:
I am happy to fix a bug anytime, but my point is, the intension of this PR is to introduce template query feature. Personally, I like to track bug and feature in different PRs, it's easier for people to read. And it will be easier if when people want to back-port bug fix to earlier version if needed. 

This approach will help keep our change history clean and make it easier to track and manage both issues independently. Please let me know if you have any questions or if you'd prefer a different approach. I'm happy to adjust as needed to align with the team's preferences.

I can create a new bug issue and open up a new PR to fix it. I can help fix it in a new PR of course if you align with this approach.

Reply:
How about the following (as a pretty simple change)?

1. Replace the `asyncActions` field with `private final AtomicReference<List<BiConsumer<Client, ActionListener<?>>>> asyncActionsRef = new AtomicReference<>(new ArrayList<>());`
2. The existing mentions of `asyncActions` (in `registerAsyncAction` and `hasAsyncActions`) can use `asyncActionsRef.get()`.
3. At the top `executeAsyncActions`, you can add a line that says ` List<BiConsumer<Client, ActionListener<?>>> asyncActions = asyncActionsRef.getAndSet(new ArrayList<>());`

Bam... no more risk of race conditions. No more need to make a copy of the array.

Reply:
Thanks @dbwiddis and @msfroh for the insights. I raised this issue in a bug ticket, and assigned myself. These fixes are also included in the suggested solution. We can continue the discussion in the bug issue https://github.com/opensearch-project/OpenSearch/issues/17103

Reply:
@dbwiddis  -- are you okay with the race condition fix as a fast-follow? 

I think we should try to get the fix in for 2.19 as well (since it's pretty easy), but I'm okay with it being a separate PR.